core/arch/riscv64/proc.c

```diff
@@ -0,0 +1,257 @@
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL RIVOS, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "../globals.h"
+#include "proc.h"
+
+static int num_simd_saved;
+static int num_simd_registers;
+static int num_opmask_registers;
+
+static size_t
+read_cache_line(const char *fname)
+{
+    char b[sizeof(uint64)], tmp;
+    size_t res = 0;
+    ssize_t n, i;
+    file_t f;
+
+    if (os_file_exists(fname, false)) {
+        f = os_open(fname, OS_OPEN_READ);
+        if (f != INVALID_FILE) {
+            n = os_read(f, b, sizeof(b));
+            if (n > 0) {
+                /* byte-swap */
+                for (i = 0; i < n / 2; i++) {
+                    tmp = b[n - i - 1];
+                    b[n - i - 1] = b[i];
+                    b[i] = tmp;
+                }
+                res = *(uint64 *)b;
+            }
+        }
+    }
+    return res;
+}
+
+/* Obtains dcache and icache line size and sets the values at the given pointers.
+ * Returns false if no value was written.
+ *
+ * Note:
+ * - Since no CSR holds those values, sysfs+device-tree is used and this works
+ *   only on Linux.
+ * - This code assumes that all harts have the same L1 cache-line size.
+ */
+static bool
+get_cache_line_size(OUT size_t *dcache_line_size, OUT size_t *icache_line_size)
+{
+#if !defined(DR_HOST_NOT_TARGET) && defined(LINUX)
+    static const char *d_cache_fname =
+        "/sys/devices/system/cpu/cpu0/of_node/d-cache-block-size";
+    static const char *i_cache_fname =
+        "/sys/devices/system/cpu/cpu0/of_node/i-cache-block-size";
+    static size_t dcache_line = 0;
+    static size_t icache_line = 0;
+    size_t sz;
+
+    if (dcache_line_size) {
+        ATOMIC_8BYTE_ALIGNED_READ(&dcache_line, &sz);
+        if (sz == 0) {
+            sz = read_cache_line(d_cache_fname);
+            if (sz > 0)
+                ATOMIC_8BYTE_ALIGNED_WRITE(&dcache_line, &sz, false);
+        }
+        ATOMIC_8BYTE_ALIGNED_READ(&dcache_line, &sz);
+        if (sz != 0)
+            *dcache_line_size = sz;
+    }
+
+    if (icache_line_size) {
+        if (icache_line == 0) {
```

Comment:
Do we need atomic reads/writes here?

Reply:
Not at all. I've missed fixing the dcache logic when I updated icache to normal writes.